### PR TITLE
chore: update Kubernetes libraries to v0.24.7

### PR DESCRIPTION
**Pull request by jackfrancis**
**Reason for Change**:

This PR updates Kubernetes libraries to `v0.24.7`, the latest v0.24 version. In addition several related other updates are appropriate to include with this change:

- Now building with golang v1.17 (from v1.16)
- Updated to the latest go-dev container image for CI jobs
- `go-autorest` updated to v0.11.18 (from v0.9.6)

Additionally, these changes brought new versions of go linting and shellcheck. The requested changes to satisfy new go linting requirements were superficial, and have been applied. The suggested changes from shellcheck are to critical areas of the bootstrap scripts, and as such my recommendation is that we ignore them, and deactivate shellcheck from CI altogether based on the deprecation/LTS goal of keeping known-working surface areas as static as possible.

**Issue Fixed**:

**Credit Where Due**:

Does this change contain code from or inspired by another project?

- [ ] No
- [ ] Yes

If "Yes," did you notify that project's maintainers and provide attribution?

- [ ] No
- [ ] Yes

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -156,7 +156,7 @@ test: generate
 	ginkgo -mod=vendor -skipPackage test/e2e -failFast -r -v -tags=fast -ldflags '$(LDFLAGS)' .
 
 .PHONY: test-style
-test-style: validate-go validate-shell validate-copyright-headers
+test-style: validate-go validate-copyright-headers
```
Comment (jackfrancis) [inline Makefile:159]: @CecileRobertMichon @mboersma this is the change that skips shellcheck

File: cmd/root_test.go [hunk 2]
```diff
@@ -216,7 +213,7 @@ func TestWriteCustomCloudProfile(t *testing.T) {
...
...//management.azurestackci15.onmicrosoft.com/36f71706-54df-4305-9847-5b038a4cf189","resourceManagerEndpoint":"https://management.local.azurestack.external/","activeDirectoryEndpoint":"https://login.windows.net/","galleryEndpoint":"https://portal.local.azurestack.external=30015/","keyVaultEndpoint":"https://vault.azurestack.external/","graphEndpoint":"https://graph.windows.net/","serviceBusEndpoint":"https://servicebus.azurestack.external/","batchManagementEndpoint":"https://batch.azurestack.external/","storageEndpointSuffix":"core.azurestack.external","sqlDatabaseDNSSuffix":"database.azurestack.external","trafficManagerDNSSuffix":"trafficmanager.cn","keyVaultDNSSuffix":"vault.azurestack.external","serviceBusEndpointSuffix":"servicebus.azurestack.external","serviceManagementVMDNSSuffix":"chinacloudapp.cn","resourceManagerVMDNSSuffix":"cloudapp.azurestack.external","containerRegistryDNSSuffix":"azurecr.io","cosmosDBDNSSuffix":"","tokenAudience":"https://management.azurestack.external/","apiManagementHostNameSuffix":"","synapseEndpointSuffix":"","resourceIdentifiers":{"graph":"","keyVault":"","datalake":"","batch":"","operationalInsights":"","storage":"","synapse":"","serviceBus":""}}`
```
Comment (jackfrancis) [inline cmd/root_test.go:216]: @jadarsie FYI updates to Azure go libs brought in some additional data here, UT have been updated.

Let me know if this is a concern.

Comment (jadarsie) [reply cmd/root_test.go:216]: No concerns that I can think of

File: pkg/engine/armvariables_test.go [hunk 3]
```diff
@@ -550,7 +550,7 @@ func TestK8sVars(t *testing.T) {
...
.../management.azurestackci15.onmicrosoft.com/36f71706-54df-4305-9847-5b038a4cf189","resourceManagerEndpoint":"https://management.local.azurestack.external/","activeDirectoryEndpoint":"https://login.windows.net/","galleryEndpoint":"https://portal.local.azurestack.external=30015/","keyVaultEndpoint":"https://vault.azurestack.external/","graphEndpoint":"https://graph.windows.net/","serviceBusEndpoint":"https://servicebus.azurestack.external/","batchManagementEndpoint":"https://batch.azurestack.external/","storageEndpointSuffix":"core.azurestack.external","sqlDatabaseDNSSuffix":"database.azurestack.external","trafficManagerDNSSuffix":"trafficmanager.cn","keyVaultDNSSuffix":"vault.azurestack.external","serviceBusEndpointSuffix":"servicebus.azurestack.external","serviceManagementVMDNSSuffix":"chinacloudapp.cn","resourceManagerVMDNSSuffix":"cloudapp.azurestack.external","containerRegistryDNSSuffix":"azurecr.io","cosmosDBDNSSuffix":"","tokenAudience":"https://management.azurestack.external/","apiManagementHostNameSuffix":"","synapseEndpointSuffix":"","resourceIdentifiers":{"graph":"","keyVault":"","datalake":"","batch":"","operationalInsights":"","storage":"","synapse":"","serviceBus":""}}`,
```
Comment (CecileRobertMichon) [inline pkg/engine/armvariables_test.go:553]: where is this diff coming from? is it expected? @jadarsie

File: scripts/validate-copyright-header.sh [hunk 4]
```diff
@@ -9,7 +9,7 @@ echo "==> Checking copyright headers <=="
 
 files=$(find . -type f -iname '*.go' ! -path './vendor/*' ! -path './hack/tools/*' ! -path './test/e2e/vendor/*')
 licRes=$(for file in $files; do
-           awk 'NR<=3' "$file" | grep -Eq "(Copyright|generated|GENERATED)" || echo "$file";
+           awk 'NR<=4' "$file" | grep -Eq "(Copyright|generated|GENERATED)" || echo "$file";
```
Comment (CecileRobertMichon) [inline scripts/validate-copyright-header.sh:12]: what is this change doing?

Comment (jackfrancis) [reply scripts/validate-copyright-header.sh:12]: golang v1.17 (probably more like gofmt) brings a new canonical build tags format that includes a newline, so there are a few files that have one more line above the grep match pattern than previously.

If I were testing your code review skills you'd get an A+, good job!

**Review by CecileRobertMichon — APPROVED** [review]
lgtm

Thank you for doing this painful but much-needed work @jackfrancis

**Pull request merged**